### PR TITLE
Adding L1MessageSender to Transaction

### PR DESCRIPTION
**Pull request by willmeister**
Adds L1MessageSender to Transaction

Fixes [YAS-377](https://optimists.atlassian.net/browse/YAS-377)

### CODE REVIEW

File: core/rawdb/accessors_indexes_test.go [hunk 1]
```diff
@@ -66,9 +66,13 @@ func TestLookupStorage(t *testing.T) {
 		t.Run(tc.name, func(t *testing.T) {
 			db := NewMemoryDatabase()
 
-			tx1 := types.NewTransaction(1, common.BytesToAddress([]byte{0x11}), big.NewInt(111), 1111, big.NewInt(11111), []byte{0x11, 0x11, 0x11})
-			tx2 := types.NewTransaction(2, common.BytesToAddress([]byte{0x22}), big.NewInt(222), 2222, big.NewInt(22222), []byte{0x22, 0x22, 0x22})
-			tx3 := types.NewTransaction(3, common.BytesToAddress([]byte{0x33}), big.NewInt(333), 3333, big.NewInt(33333), []byte{0x33, 0x33, 0x33})
+			sender1 := common.BytesToAddress([]byte{0x44})
```
Comment (ben-chain) [inline core/rawdb/accessors_indexes_test.go:69]: To confirm--this is testing basically that the transactions are correctly stored in the DB right?  If so, might be good to test those with and without an L1MsgSender.   (maybe I'm jumping the gun and those exist below)

File: core/types/transaction.go [hunk 2]
```diff
@@ -71,28 +72,35 @@ type txdataMarshaling struct {
...
 
-func NewTransaction(nonce uint64, to common.Address, amount *big.Int, gasLimit uint64, gasPrice *big.Int, data []byte) *Transaction {
-	return newTransaction(nonce, &to, amount, gasLimit, gasPrice, data)
+func NewTransaction(nonce uint64, to common.Address, amount *big.Int, gasLimit uint64, gasPrice *big.Int, data []byte, l1MessageSender *common.Address) *Transaction {
+	if l1MessageSender == nil {
+		return newTransaction(nonce, &to, amount, gasLimit, gasPrice, data, nil)
+	} else {
+		sender := new(common.Address)
+		*sender = *l1MessageSender
+		return newTransaction(nonce, &to, amount, gasLimit, gasPrice, data, sender)
```
Comment (ben-chain) [inline core/types/transaction.go]: Just for my own learning, why is this logic necessary?  Seems like you want to be sure that the `sender` passed into `newTransaction` is at a different spot in memory/unique right?  Why is that needed here and not for the other. inputs like `amount`?

Comment (ben-chain) [reply core/types/transaction.go]: Or maybe `amount` is easier as it's a built in type so I guess a better question is why not pass `&l1MessageSender` like with `&to`?

File: core/types/transaction.go [hunk 3]
```diff
@@ -189,13 +197,33 @@ func (tx *Transaction) To() *common.Address {
 	return &to
 }
 
+// L1MessageSender returns the L1 message sender address of the transaction if one exists.
+// It returns nil if this transaction was not from an L1 contract..
```
Comment (ben-chain) [inline core/types/transaction.go]:
```suggestion
// It returns nil if this transaction was not from an L1 contract.
```

File: core/types/transaction.go [hunk 4]
```diff
@@ -386,17 +415,18 @@ func (t *TransactionsByPriceAndNonce) Pop() {
...
+	nonce           uint64
+	amount          *big.Int
+	gasLimit        uint64
+	gasPrice        *big.Int
+	data            []byte
+	checkNonce      bool
 }
 
-func NewMessage(from common.Address, to *common.Address, nonce uint64, amount *big.Int, gasLimit uint64, gasPrice *big.Int, data []byte, checkNonce bool) Message {
+func NewMessage(from common.Address, to *common.Address, nonce uint64, amount *big.Int, gasLimit uint64, gasPrice *big.Int, data []byte, checkNonce bool, l1MessageSender *common.Address) Message {
```
Comment (ben-chain) [inline core/types/transaction.go:423]: EDIT: I think the below must be wrong, ignore!

> I'm assuming these are EVM internal messages e.g. generated by `CALL` opcode,  right?  If so, one note is that this will never change over the course of the transaction so it could be handled however tx.origin is handled.

File: core/types/transaction.go [hunk 5]
```diff
@@ -219,13 +247,14 @@ func (tx *Transaction) Size() common.StorageSize {
 // XXX Rename message to something less arbitrary?
```
Comment (ben-chain) [inline core/types/transaction.go:241]: Below you'll notice I got confused about what these `message`s are.  Since there is a note here already in geth to consider renaming this, and we are adding a new "message" keyword in this PR, we might want to consider it. xD

File: core/types/transaction_test.go [hunk 6]
```diff
@@ -70,6 +65,14 @@ func TestTransactionEncode(t *testing.T) {
 	if !bytes.Equal(txb, should) {
 		t.Errorf("encoded RLP mismatch, got %x", txb)
 	}
+
+	txc, err := rlp.EncodeToBytes(rightvrsTxWithL1Sender)
+	if err != nil {
+		t.Fatalf("encode error: %v", err)
+	}
+	if bytes.Equal(txc, should) {
+		t.Errorf("RLP encoding with L1MessageSender should be different than without. Got %x", txc)
```
Comment (ben-chain) [inline core/types/transaction_test.go:74]: 👌👌👌

File: core/types/transaction_test.go [hunk 7]
```diff
@@ -217,5 +220,19 @@ func TestTransactionJSON(t *testing.T) {
 		if tx.ChainId().Cmp(parsedTx.ChainId()) != 0 {
 			t.Errorf("invalid chain id, want %d, got %d", tx.ChainId(), parsedTx.ChainId())
 		}
+		if tx.L1MessageSender() != nil && parsedTx.L1MessageSender() != nil && *tx.L1MessageSender() != *parsedTx.L1MessageSender() {
```
Comment (ben-chain) [inline core/types/transaction_test.go]: Noting that it seems like we would not catch an error here if `tx.L1MessageSender() != nil` but `parsedTx.L1MessageSender() == nil`.  Not positive it's an issue but thought worth noting.

File: core/types/transaction_test.go [hunk 8]
```diff
@@ -217,5 +220,19 @@ func TestTransactionJSON(t *testing.T) {
...
+			t.Errorf("invalid L1MessageSender, want %x, got %x", tx.L1MessageSender(), parsedTx.L1MessageSender())
+		}
+	}
+}
+
+// Tests that L1MessageSender has no impact on hash
+func TestL1MessageSenderHash(t *testing.T) {
```
Comment (ben-chain) [inline core/types/transaction_test.go:230]: I see the logic above where you are enforcing this, but don't fully understand why--is it related to not breaking tests?

I feel like it shouldn't matter either way since these transactions will already be hashed on chain, and probably not with an RLP encoding.  But still confused why this is done.

File: eth/helper_test.go [hunk 9]
```diff
@@ -71,9 +71,9 @@ func newTestProtocolManager(mode downloader.SyncMode, blocks int, generator func
 		panic(err)
 	}
 	blockSubmitter := rollup.NewBlockSubmitter()
-	rollupBlockBuilder, err := rollup.NewTransitionBatchBuilder(db, blockchain, blockSubmitter, 5 * time.Minute, 9_000_000_000, 200)
+	rollupBlockBuilder, err := rollup.NewTransitionBatchBuilder(db, blockchain, blockSubmitter, 5*time.Minute, 9_000_000_000, 200)
 	if err != nil {
-		panic(fmt.Errorf("failed to create Rollup Block Builder: %v", err)
+		panic(fmt.Errorf("failed to create Rollup Block Builder: %v", err))
```
Comment (ben-chain) [inline eth/helper_test.go:76]: Oh wow, were our builds previously failing? 😅😂

**Review by ben-chain — APPROVED** [review]
Woohoo!! This is looking awesome, super sweet to have the whole `?` addition to the RLP--looks super clean 😀

I left a few small comments--I think most critical question is understanding why this new field is being omitted from the hash even if present as it feels scary to have equivalent hashes for two different transactions.

Last thought--not sure it's necessarily in this issue's scope, but we will need to make sure that users cannot `sendRawTransaction` with an `L1MessageSender` as this would result in accidental fraud.  Might be another ticket, but hadn't thought of it until just now while reviewing.

File: core/rawdb/accessors_indexes_test.go (hunk 1 above)
Comment (willmeister) [reply core/rawdb/accessors_indexes_test.go:69]: Yep, couldn't hurt To leave one of them `nil`

File: core/types/transaction.go (hunk 2 above)
Comment (willmeister) [reply core/types/transaction.go]: Yeah, don't want modifications to vars that should be immutable from outside of this package. Can put the burden on all the different callers or just do it here and know for sure. If done that way everywhere else, though, I can update here to follow convention.

File: core/types/transaction.go (hunk 5 above)
Comment (willmeister) [reply core/types/transaction.go:241]: Not sure how this is used, just updated it since it is Transaction-like. Even if poorly named, we should probably keep parity with Geth (👀) on anything that we don't have a good reason to change or pulling in updates from Geth will be even hairier than it would otherwise be.

File: core/types/transaction_test.go (hunk 7 above)
Comment (willmeister) [reply core/types/transaction_test.go]: If `parsedTx.L1MessageSender() == nil`, evaluating `*parsedTx.L1MessageSender()` will throw a Null Pointer Exception, since we're trying to access the memory address pointed to by `nil`.

File: core/types/transaction_test.go (hunk 8 above)
Comment (willmeister) [reply core/types/transaction_test.go:230]: Yeah, this would break all tests with pre-computed hashes and interaction with any 3rd party tools (for instance, ethers throws saying that hash != expected hash). I'd be concerned if it's possible to have the same exact message but with a different L1 sender or an empty one, but the L1 sender will only be populated and submitted by our L1 subscriber, so the nonce, at a minimum, should be different, right? If so, then we can maintain compatibility without worrying about uniqueness.

File: eth/helper_test.go (hunk 9 above)
Comment (willmeister) [reply eth/helper_test.go:76]: What's weird is that builds were not, but yes, running tests would fail. It appears as if tests are JIT compiled or something.

File: core/types/transaction_test.go (hunk 7 above)
Comment (ben-chain) [reply core/types/transaction_test.go]: Ahh great, and presumably all three args are evaluated even if the second one is false.   Got it!

File: core/types/transaction.go (hunk 2 above)
Comment (willmeister) [reply core/types/transaction.go]: I'll update it to make the caller choose. Caller can copy safely ahead of time if they think it makes sense.

File: core/types/transaction_test.go (hunk 8 above)
Comment (ben-chain) [reply core/types/transaction_test.go:230]: Ahhh okay yeah I see your point WRT compatibility.

Hadn't thought of nonces being different too, so good point there, I think you're right.  It would require that we always pass in the correct nonce in these transactions, which shouldn't be a problem.  Although it's. an interesting point that we don't currently expose that nonce to the execution context unless I'm mistaken.  Might want to add that 🤔

File: core/types/transaction_test.go (hunk 7 above)
Comment (willmeister) [reply core/types/transaction_test.go]: No, it should eagerly exit the evaluation. Sorry, I misread your comment. Yes, I'll add a check if one is nil but the other is not.
